src/function/scalar/list/list_lambdas.cpp

```diff
@@ -0,0 +1,375 @@
+			result_entries[row_idx].length = list_entry.length;
+			offset += list_entry.length;
+		} else {
+			lists_len.push_back(list_entry.length);
+		}
+
+		// empty list, nothing to execute
+		if (list_entry.length == 0) {
+			continue;
+		}
+
+		// get the data indexes
+		for (idx_t col_idx = 0; col_idx < args.ColumnCount() - 1; col_idx++) {
+			indexes[col_idx] = columns[col_idx].sel->get_index(row_idx);
+		}
+
+		// iterate list elements and create transformed expression columns
+		for (idx_t child_idx = 0; child_idx < list_entry.length; child_idx++) {
+
+			// reached STANDARD_VECTOR_SIZE elements
+			if (elem_cnt == STANDARD_VECTOR_SIZE) {
+
+				lambda_chunk.Reset();
+				ExecuteExpression(types, result_types, elem_cnt, sel, sel_vectors, input_chunk, lambda_chunk,
+				                  child_vector, args, expr_executor);
+
+				auto &lambda_vector = lambda_chunk.data[0];
+
+				if (IS_TRANSFORM) {
+					AppendTransformedToResult(lambda_vector, elem_cnt, result);
+				} else {
+					AppendFilteredToResult(lambda_vector, result_entries, elem_cnt, result, curr_list_len,
+					                       curr_list_offset, appended_lists_cnt, lists_len, curr_original_list_len,
+					                       input_chunk);
+				}
+				elem_cnt = 0;
+			}
+
+			// to slice the child vector
+			auto source_idx = child_data.sel->get_index(list_entry.offset + child_idx);
+			sel.set_index(elem_cnt, source_idx);
+
+			// for each column, set the index of the selection vector to slice properly
+			for (idx_t col_idx = 0; col_idx < args.ColumnCount() - 1; col_idx++) {
+				sel_vectors[col_idx].set_index(elem_cnt, indexes[col_idx]);
+			}
+			elem_cnt++;
+		}
+	}
+
+	lambda_chunk.Reset();
+	ExecuteExpression(types, result_types, elem_cnt, sel, sel_vectors, input_chunk, lambda_chunk, child_vector, args,
+	                  expr_executor);
+	auto &lambda_vector = lambda_chunk.data[0];
+
+	if (IS_TRANSFORM) {
+		AppendTransformedToResult(lambda_vector, elem_cnt, result);
+	} else {
+		AppendFilteredToResult(lambda_vector, result_entries, elem_cnt, result, curr_list_len, curr_list_offset,
+		                       appended_lists_cnt, lists_len, curr_original_list_len, input_chunk);
+	}
+}
+
+static void ListTransformFunction(DataChunk &args, ExpressionState &state, Vector &result) {
+	ListLambdaFunction<>(args, state, result);
+}
+
+static void ListFilterFunction(DataChunk &args, ExpressionState &state, Vector &result) {
+	ListLambdaFunction<false>(args, state, result);
+}
+
+template <int64_t LAMBDA_PARAM_CNT>
+static unique_ptr<FunctionData> ListLambdaBind(ClientContext &context, ScalarFunction &bound_function,
+                                               vector<unique_ptr<Expression>> &arguments) {
+
+	// this is pretty ugly, but I did not come up with a better way to ensure the correct number of
+	// lhs lambda parameters, and to ensure that the lambda expression is not NULL
+	idx_t num_params = 0;
+	if (!arguments[arguments.size() - 1]->alias.empty()) {
```

Comment:
Can't you look at the lambda expression directly? Using an alias for this is quite hacky.

Reply:
Only before the transformation/iteration of the expression, which happens in the `BindLambdaFunction`. In the specific binds (`ListTransformBind`, `ListFilterBind`) that information is no longer in the lambda expression.
E.g. this is how the expression looks like. 
```
(std::string) $3 = "(CAST(#0 AS INTEGER) + #1)"
```
Now, I cannot just count all of the `#0` references, because ALL different lambda parameters will reference the first column/list input at this point.

So I need a way to get that parameter count from the `BindLambdaFunction` to the specific binds... and this was the only very hacky idea that I had...

Reply:
We need the transformed lambda expression for `ScalarFunction::Bind`. We move it into the `ListLambdaBindData` so that we can create the expression executor with the lambda expression.
```c++
return make_unique<ListLambdaBindData>(bound_function.return_type, move(lambda_expr));
```